tao_compiler/mlir/disc/utils/source_emitter.cc

```diff
@@ -0,0 +1,601 @@
+  if (!bcast) {
+    return llvm::None;
+  }
+
+  // Only deal with the case that the last rewriter in the same block is the
+  // ConstantOp.
+  auto input_value = op->getOperand(0);
+  auto input_op = findLastWriterInBlock(input_value, op->getBlock());
+  if (!input_op.hasValue()) {
+    return llvm::None;
+  }
+
+  lmhlo::ConstantOp constant =
+      dyn_cast_or_null<lmhlo::ConstantOp>(input_op.value());
+  if (!input_op) {
+    return llvm::None;
+  }
+
+  if (binding.count(input_value) == 0) {
+    llvm::None;
+  }
+  std::string expression = binding[input_value];
+
+  Value result = op->getOperand(2);
+  MemRefType memref_type = result.getType().cast<MemRefType>();
+  Type result_type = memref_type.getElementType();
+  std::string type_str = MLIRType2CUDATypeStr(result_type);
+
+  std::string result_name = EmitUniqueName("dyn_bcast_indim");
+
+  assert(binding.count(result) == 0);
+  binding[result] = result_name;
+
+  return type_str + " " + result_name + " = " + expression;
+}
+
+llvm::Optional<std::string> SourceEmitterCUDA::EmitOp(
+    Operation* op, ValueNameBinding& binding) {
+  if (isa<lmhlo::AbsOp, lmhlo::CeilOp, lmhlo::ConvertOp, lmhlo::CosineOp,
+          lmhlo::ExpOp, lmhlo::FloorOp, lmhlo::IsFiniteOp, lmhlo::LogOp,
+          lmhlo::Log1pOp, lmhlo::NegOp, lmhlo::NotOp, lmhlo::RsqrtOp,
+          lmhlo::SqrtOp, lmhlo::TanhOp>(op)) {
+    return EmitElemWiseUnaryOp(op, binding);
+  } else if (isa<lmhlo::AddOp, lmhlo::SubtractOp, lmhlo::MulOp, lmhlo::DivOp,
+                 lmhlo::MaxOp, lmhlo::MinOp, lmhlo::CompareOp, lmhlo::AndOp,
+                 lmhlo::OrOp, lmhlo::RemOp, lmhlo::PowOp>(op)) {
+    return EmitElemWiseBinaryOp(op, binding);
+  } else if (isa<lmhlo::SelectOp, lmhlo::ClampOp>(op)) {
+    return EmitElemWiseTernaryOp(op, binding);
+  } else if (isa<lmhlo::ConstantOp>(op)) {
+    return EmitScalarOrSplatConstantOp(op, binding);
+  } else if (isa<lmhlo::DynamicBroadcastInDimOp>(op)) {
+    return EmitBroadcastOfScalarOrSplatConstantOp(op, binding);
+  } else {
+    return llvm::None;
+  }
+}
+
+std::string SourceEmitterCUDA::EmitUniqueName(llvm::StringRef op_str) {
+  std::string name = "bladedisc_" + op_str.str();
+  if (existing_names_.count(op_str.str()) == 0) {
+    existing_names_.try_emplace(op_str.str(), 0);
+  }
+  int32_t count = existing_names_[op_str.str()]++;
+  name += "_" + to_string(count);
+
+  return name;
+}
+
+void SourceEmitterCUDA::bindValueNames(
+    const SmallVectorImpl<Value>& inputs,
+    const SmallVectorImpl<std::string>& names, ValueNameBinding& binding) {
+  assert(inputs.size() == names.size());
+  for (int64_t i = 0; i < inputs.size(); i++) {
+    binding[inputs[i]] = names[i];
+  }
+}
+
+bool SourceEmitterCUDA::isSupportedOp(Operation* op) {
+  if (isa<lmhlo::AbsOp, lmhlo::CeilOp, lmhlo::ConvertOp, lmhlo::CosineOp,
+          lmhlo::ExpOp, lmhlo::FloorOp, lmhlo::IsFiniteOp, lmhlo::LogOp,
+          lmhlo::Log1pOp, lmhlo::NegOp, lmhlo::NotOp, lmhlo::RsqrtOp,
```

Comment:
Some quantization related op may also need to be added such as 'lmhlo.round_nearest_even'. I will try to extend supported ops after the basic functions can be run through.

Reply:
Sure.